File: rust-functions/common/src/clients.rs
//! AWS SDK client initialization with lazy loading and endpoint overrides.

use aws_config::BehaviorVersion;
use std::sync::OnceLock;

/// Global DynamoDB client instance.
static DYNAMODB_CLIENT: OnceLock<aws_sdk_dynamodb::Client> = OnceLock::new();

/// Global S3 client instance.
static S3_CLIENT: OnceLock<aws_sdk_s3::Client> = OnceLock::new();

/// Global Cognito Identity Provider client instance.
static COGNITO_CLIENT: OnceLock<aws_sdk_cognitoidentityprovider::Client> = OnceLock::new();

/// Global CloudWatch client instance.
static CLOUDWATCH_CLIENT: OnceLock<aws_sdk_cloudwatch::Client> = OnceLock::new();

/// Returns a shared DynamoDB client instance.
///
/// The client is lazily initialized on first call and reused for subsequent calls.
/// Supports custom endpoint override via `DYNAMODB_ENDPOINT` environment variable
/// for local testing with DynamoDB Local or LocalStack.
pub async fn get_dynamodb_client() -> &'static aws_sdk_dynamodb::Client {
    DYNAMODB_CLIENT.get_or_init(|| {
        tokio::runtime::Handle::current().block_on(async {
            let config = load_aws_config().await;
            aws_sdk_dynamodb::Client::new(&config)
        })
    })
}

/// Returns a shared S3 client instance.
///
/// The client is lazily initialized on first call and reused for subsequent calls.
/// Supports custom endpoint override via `S3_ENDPOINT` environment variable
/// for local testing with LocalStack.
pub async fn get_s3_client() -> &'static aws_sdk_s3::Client {
    S3_CLIENT.get_or_init(|| {
        tokio::runtime::Handle::current().block_on(async {
            let config = load_aws_config().await;
            aws_sdk_s3::Client::new(&config)
        })
    })
}

/// Returns a shared Cognito Identity Provider client instance.
///
/// The client is lazily initialized on first call and reused for subsequent calls.
/// Supports custom endpoint override via `COGNITO_ENDPOINT` environment variable
/// for local testing with LocalStack.
pub async fn get_cognito_client() -> &'static aws_sdk_cognitoidentityprovider::Client {
    COGNITO_CLIENT.get_or_init(|| {
        tokio::runtime::Handle::current().block_on(async {
            let config = load_aws_config().await;
            aws_sdk_cognitoidentityprovider::Client::new(&config)
        })
    })
}

/// Returns a shared CloudWatch client instance.
///
/// The client is lazily initialized on first call and reused for subsequent calls.
pub async fn get_cloudwatch_client() -> &'static aws_sdk_cloudwatch::Client {
    CLOUDWATCH_CLIENT.get_or_init(|| {
        tokio::runtime::Handle::current().block_on(async {
            let config = load_aws_config().await;
            aws_sdk_cloudwatch::Client::new(&config)
        })
    })
}

/// Loads AWS configuration with optional endpoint override.
async fn load_aws_config() -> aws_config::SdkConfig {
    let mut config_loader = aws_config::defaults(BehaviorVersion::latest());

    // Check for LocalStack/DynamoDB Local endpoint override
    if let Ok(endpoint) = std::env::var("AWS_ENDPOINT_URL") {
        config_loader = config_loader.endpoint_url(&endpoint);
    }

    config_loader.load().await
}

/// Resets all cached clients. Only for testing purposes.
#[cfg(test)]
pub fn reset_clients() {
    // OnceLock doesn't support resetting, so this is a no-op
    // In tests, each test should use fresh clients if needed
}

#[cfg(test)]
mod tests {
    // Client initialization tests require AWS credentials
    // They are covered in integration tests
}

File: go-functions/internal/clients/clients.go
// Package clients provides singleton AWS service client initialization.
// It supports thread-safe lazy initialization using sync.Once,
// environment variable configuration for region, and LocalStack endpoint overrides.
package clients

import (
	"context"
	"os"
	"sync"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

var (
	dynamoClient  *dynamodb.Client
	s3Client      *s3.Client
	cognitoClient *cognitoidentityprovider.Client
	presignClient *s3.PresignClient
	once          sync.Once
	initErr       error
)

// initClients initializes all AWS clients once.
// It reads region from AWS_REGION environment variable and
// supports LocalStack endpoint overrides via DYNAMODB_ENDPOINT,
// S3_ENDPOINT, and COGNITO_ENDPOINT environment variables.
func initClients() {
	cfg, err := config.LoadDefaultConfig(context.Background())
	if err != nil {
		initErr = err
		return
	}

	// Initialize DynamoDB client with optional endpoint override
	dynamoOpts := []func(*dynamodb.Options){}
	if endpoint := os.Getenv("DYNAMODB_ENDPOINT"); endpoint != "" {
		dynamoOpts = append(dynamoOpts, func(o *dynamodb.Options) {
			o.BaseEndpoint = aws.String(endpoint)
		})
	}
	dynamoClient = dynamodb.NewFromConfig(cfg, dynamoOpts...)

	// Initialize S3 client with optional endpoint override
	s3Opts := []func(*s3.Options){}
	if endpoint := os.Getenv("S3_ENDPOINT"); endpoint != "" {
		s3Opts = append(s3Opts, func(o *s3.Options) {
			o.BaseEndpoint = aws.String(endpoint)
			o.UsePathStyle = true // Required for LocalStack
		})
	}
	s3Client = s3.NewFromConfig(cfg, s3Opts...)

	// Initialize S3 Presign client
	presignClient = s3.NewPresignClient(s3Client)

	// Initialize Cognito client with optional endpoint override
	cognitoOpts := []func(*cognitoidentityprovider.Options){}
	if endpoint := os.Getenv("COGNITO_ENDPOINT"); endpoint != "" {
		cognitoOpts = append(cognitoOpts, func(o *cognitoidentityprovider.Options) {
			o.BaseEndpoint = aws.String(endpoint)
		})
	}
	cognitoClient = cognitoidentityprovider.NewFromConfig(cfg, cognitoOpts...)
}

// GetDynamoDB returns the singleton DynamoDB client.
func GetDynamoDB() (*dynamodb.Client, error) {
	once.Do(initClients)
	if initErr != nil {
		return nil, initErr
	}
	return dynamoClient, nil
}

// GetS3 returns the singleton S3 client.
func GetS3() (*s3.Client, error) {
	once.Do(initClients)
	if initErr != nil {
		return nil, initErr
	}
	return s3Client, nil
}

// GetCognito returns the singleton Cognito Identity Provider client.
func GetCognito() (*cognitoidentityprovider.Client, error) {
	once.Do(initClients)
	if initErr != nil {
		return nil, initErr
	}
	return cognitoClient, nil
}

// GetPresignClient returns the singleton S3 Presign client for generating presigned URLs.
func GetPresignClient() (*s3.PresignClient, error) {
	once.Do(initClients)
	if initErr != nil {
		return nil, initErr
	}
	return presignClient, nil
}

// ResetForTesting resets the singleton state for testing purposes.
// This should only be used in tests.
func ResetForTesting() {
	once = sync.Once{}
	dynamoClient = nil
	s3Client = nil
	cognitoClient = nil
	presignClient = nil
	initErr = nil
}

// SetInitErrorForTesting sets an initialization error for testing error paths.
// This should only be used in tests.
func SetInitErrorForTesting(err error) {
	once.Do(func() {
		initErr = err
	})
}
